### ISSUE TITLE
False positive: Color-contrast fails on a gradient with pseudo selector

### ISSUE THREAD

**Issue by iamrafan** [dequelabs/axe-core#975]
This issue should show up as needs review, like cases where gradient is used on non pseudo selectors

- Link to live demo: https://codepen.io/anon/pen/pKOmwy

<pre><code>axe-core version: 3.0.3
axe-Coconut version: 3.4.0-alpha.2</code></pre>

**Comment by marcysutton** [comment]
@iamrafan can you put the demo on Codepen or similar so we don't have to download a zip file (for both security and convenience purposes)

**Comment by marcysutton** [comment]
axe does send gradient backgrounds into needs review, but it doesn't yet support background contrast testing on CSS pseudo elements or borders. We've seen a few issues before with both pseudos and borders, but not in high enough numbers to warrant a fix. 

This could put in more general terms: axe should test pseudo element backgrounds (regardless of a gradient). I added some related comments in #643: we'd have to query window.getComputedStyle for the element, then look at its `:before` and `:after` content, and see if it overlaps.

**Comment by iamrafan** [comment]
@marcysutton I updated the demo link

**Comment by marcysutton** [comment]
This issue is on hold until we figure out whether the CSSOM stuff being done is actually going to work. Then it will be an issue of calculating the position of pseudo elements to see if they overlap other content as these elements don't end up in `document.elementsFromPoint` stacks. It's a lot of work for what is admittedly somewhat of an edge case.

**Comment by WilcoFiers** [comment]
I feel your pain Marcy - but it's a false positive, so we're committed to fixing it.

**Comment by marcysutton** [comment]
To update this issue with what we've found so far:

- There is a work-in-progress PR in #1220, where I did some of the experimentation to see how we might accomplish this rule.
- Pseudo elements don't have bounding boxes as they are virtual nodes, so to leverage DOM APIs like `elementsFromPoint` to check overlap we either need to do heavy calculations or append real DOM nodes with the styles gathered from pseudo elements.
- We decided to try gathering relevant `:before` and `:after` selectors with CSSOM, because querying a node's before and after pseudo elements with `window.getComputedStyle` return styles regardless of whether they have the `content` property (including a transparent black background color)
- Once you have a list of pseudo elements, the approach we discussed was to inject a DIV for each element and copy the styles from the pseudo ones. We also talked about storing bounding box & position information, storing in an arbitrary object structure for later use, and deleting the injected nodes.

Potential issues:
- Injecting DOM nodes could negatively impact CSS styles, including things like `nth-child` selectors and positioning
- Injecting DOM nodes could trigger unwanted MutationObserver events 
- Removing the created DOM nodes before running the supporting methods for `getBackgroundColor` could limit which DOM APIs can be used (e.g. if they require actual DOM nodes instead of bounding box coordinates)

**Comment by AutoSponge** [comment]
@jkodu I have some code we use for pseudo elements in image detection that could help.

**Comment by iamrafan** [comment]
'Color-contrast' also fails for text having a background image added using pseudo selector.
Here is an example: https://codepen.io/iamrafan/pen/zYGbmjp

**Comment by WilcoFiers** [comment]
Super rough drafty code on how we can do this:

```js
	while (parent = parent.parentElement) {
		const parentStyle = window.getComputedStyle(parent)
		if (['inline', 'inline-block'].includes(parentStyle.getPropertyValue('display'))) {
			break;
		}
		const beforeStyle = window.getComputedStyle(parent, 'before') // Also "after"
		if (
			hasProperty(beforeStyle, 'background') &&
			hasProperty(beforeStyle, 'position', 'absolute') &&
			hasDimensionsNotZero(beforeStyle)
		) {
			return undefined
		}
	}
```

**Comment by padmavemulapati** [comment]
Verified with the latest axe-core-develop code (dated:6th Nov 2020), 
As the issue is false positive and the fix is done for pseudo content  , tested scenarios
1. when  pseudo element has a background color: results showing "incomplete"
![image](https://user-images.githubusercontent.com/47409579/98377113-6949d280-206a-11eb-8ea9-dfb85a8a7bc4.png)
2. when pseudo element has no content, results showing "pass" for color-contrast.
![image](https://user-images.githubusercontent.com/47409579/98377444-cfcef080-206a-11eb-9a7b-49faa1d4fc51.png)
